### ISSUE TITLE
Building PeerTube CLI

### ISSUE THREAD

**Issue by faust64** [Chocobozzz/PeerTube#3221]
I'm trying to install PeerTube CLI -- more generally, to install and configure the ldap authentication plugin.

```
# NODE_ENV=build NOCLIENT=1 yarn install --pure-lockfile
# NOCLIENT=1 npm run setup:cli

> peertube@2.4.0 setup:cli /app
> scripty

yarn install v1.22.4
[1/5] Validating package.json...
[2/5] Resolving packages...
warning Resolution field "http-signature@1.3.4" is incompatible with requested version "http-signature@~1.2.0"
warning Resolution field "oauth2-server@3.1.0-beta.1" is incompatible with requested version "oauth2-server@3.0.0"
[3/5] Fetching packages...
info fsevents@2.1.3: The platform "linux" is incompatible with this module.
info "fsevents@2.1.3" is an optional dependency and failed compatibility check. Excluding it from installation.
[4/5] Linking dependencies...
warning " > @typescript-eslint/eslint-plugin@3.10.0" has unmet peer dependency "@typescript-eslint/parser@^3.0.0".
warning " > marked-man@0.7.0" has incorrect peer dependency "marked@^0.7.0".
[5/5] Building fresh packages...
$ test -n "$NOCLIENT" || (cd client && yarn install --pure-lockfile)
Done in 4.71s.
yarn install v1.22.4
[1/4] Resolving packages...
success Already up-to-date.
Done in 0.30s.

> peertube@2.4.0 tsc /app
> tsc "--build" "./server/tools/tsconfig.json"

shared/extra-utils/miscs/miscs.ts:3:23 - error TS2307: Cannot find module 'chai' or its corresponding type declarations.

3 import * as chai from 'chai'
                        ~~~~~~

shared/extra-utils/miscs/miscs.ts:5:26 - error TS2307: Cannot find module 'supertest' or its corresponding type declarations.

5 import * as request from 'supertest'
                           ~~~~~~~~~~~

shared/extra-utils/requests/requests.ts:3:26 - error TS2307: Cannot find module 'supertest' or its corresponding type declarations.

3 import * as request from 'supertest'
                           ~~~~~~~~~~~

shared/extra-utils/server/servers.ts:7:24 - error TS2307: Cannot find module 'chai' or its corresponding type declarations.

7 import { expect } from 'chai'
                         ~~~~~~

shared/extra-utils/server/clients.ts:1:26 - error TS2307: Cannot find module 'supertest' or its corresponding type declarations.

1 import * as request from 'supertest'
                           ~~~~~~~~~~~

shared/extra-utils/users/login.ts:1:26 - error TS2307: Cannot find module 'supertest' or its corresponding type declarations.

1 import * as request from 'supertest'
                           ~~~~~~~~~~~

shared/extra-utils/users/users.ts:1:26 - error TS2307: Cannot find module 'supertest' or its corresponding type declarations.

1 import * as request from 'supertest'
                           ~~~~~~~~~~~

shared/extra-utils/videos/video-channels.ts:3:26 - error TS2307: Cannot find module 'supertest' or its corresponding type declarations.

3 import * as request from 'supertest'
                           ~~~~~~~~~~~

server/helpers/custom-validators/misc.ts:86:35 - error TS2503: Cannot find namespace 'Express'.

86   files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[],
                                     ~~~~~~~

server/helpers/custom-validators/misc.ts:86:61 - error TS2503: Cannot find namespace 'Express'.

86   files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[],
                                                               ~~~~~~~

server/helpers/custom-validators/users.ts:104:56 - error TS2503: Cannot find namespace 'Express'.

104 function isAvatarFile (files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[]) {
                                                           ~~~~~~~

server/helpers/custom-validators/users.ts:104:82 - error TS2503: Cannot find namespace 'Express'.

104 function isAvatarFile (files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[]) {
                                                                                     ~~~~~~~

shared/extra-utils/server/activitypub.ts:1:26 - error TS2307: Cannot find module 'supertest' or its corresponding type declarations.

1 import * as request from 'supertest'
                           ~~~~~~~~~~~

shared/extra-utils/server/jobs.ts:1:26 - error TS2307: Cannot find module 'supertest' or its corresponding type declarations.

1 import * as request from 'supertest'
                           ~~~~~~~~~~~

shared/extra-utils/server/follows.ts:1:26 - error TS2307: Cannot find module 'supertest' or its corresponding type declarations.

1 import * as request from 'supertest'
                           ~~~~~~~~~~~

shared/extra-utils/videos/services.ts:1:26 - error TS2307: Cannot find module 'supertest' or its corresponding type declarations.

1 import * as request from 'supertest'
                           ~~~~~~~~~~~

shared/extra-utils/videos/videos.ts:3:24 - error TS2307: Cannot find module 'chai' or its corresponding type declarations.

3 import { expect } from 'chai'
                         ~~~~~~

shared/extra-utils/videos/videos.ts:7:26 - error TS2307: Cannot find module 'supertest' or its corresponding type declarations.

7 import * as request from 'supertest'
                           ~~~~~~~~~~~

shared/extra-utils/videos/video-playlists.ts:11:24 - error TS2307: Cannot find module 'chai' or its corresponding type declarations.

11 import { expect } from 'chai'
                          ~~~~~~

shared/extra-utils/users/accounts.ts:3:26 - error TS2307: Cannot find module 'supertest' or its corresponding type declarations.

3 import * as request from 'supertest'
                           ~~~~~~~~~~~

shared/extra-utils/users/accounts.ts:4:24 - error TS2307: Cannot find module 'chai' or its corresponding type declarations.

4 import { expect } from 'chai'
                         ~~~~~~

shared/extra-utils/videos/video-abuses.ts:1:26 - error TS2307: Cannot find module 'supertest' or its corresponding type declarations.

1 import * as request from 'supertest'
                           ~~~~~~~~~~~

shared/extra-utils/videos/video-blacklist.ts:1:26 - error TS2307: Cannot find module 'supertest' or its corresponding type declarations.

1 import * as request from 'supertest'
                           ~~~~~~~~~~~

shared/extra-utils/videos/video-captions.ts:2:26 - error TS2307: Cannot find module 'supertest' or its corresponding type declarations.

2 import * as request from 'supertest'
                           ~~~~~~~~~~~

shared/extra-utils/videos/video-captions.ts:3:23 - error TS2307: Cannot find module 'chai' or its corresponding type declarations.

3 import * as chai from 'chai'
                        ~~~~~~

shared/extra-utils/videos/video-comments.ts:3:26 - error TS2307: Cannot find module 'supertest' or its corresponding type declarations.

3 import * as request from 'supertest'
                           ~~~~~~~~~~~

shared/extra-utils/videos/video-streaming-playlists.ts:4:24 - error TS2307: Cannot find module 'chai' or its corresponding type declarations.

4 import { expect } from 'chai'
                         ~~~~~~

shared/extra-utils/videos/video-change-ownership.ts:1:26 - error TS2307: Cannot find module 'supertest' or its corresponding type declarations.

1 import * as request from 'supertest'
                           ~~~~~~~~~~~

shared/extra-utils/feeds/feeds.ts:1:26 - error TS2307: Cannot find module 'supertest' or its corresponding type declarations.

1 import * as request from 'supertest'
                           ~~~~~~~~~~~

shared/extra-utils/search/videos.ts:3:26 - error TS2307: Cannot find module 'supertest' or its corresponding type declarations.

3 import * as request from 'supertest'
                           ~~~~~~~~~~~

server/helpers/custom-validators/videos.ts:83:55 - error TS2503: Cannot find namespace 'Express'.

83 function isVideoFile (files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[]) {
                                                         ~~~~~~~

server/helpers/custom-validators/videos.ts:83:81 - error TS2503: Cannot find namespace 'Express'.

83 function isVideoFile (files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[]) {
                                                                                   ~~~~~~~

server/helpers/custom-validators/videos.ts:92:56 - error TS2503: Cannot find namespace 'Express'.

92 function isVideoImage (files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[], field: string) {
                                                          ~~~~~~~

server/helpers/custom-validators/videos.ts:92:82 - error TS2503: Cannot find namespace 'Express'.

92 function isVideoImage (files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[], field: string) {
                                                                                    ~~~~~~~

server/helpers/custom-validators/video-captions.ts:12:62 - error TS2503: Cannot find namespace 'Express'.

12 function isVideoCaptionFile (files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[], field: string) {
                                                                ~~~~~~~

server/helpers/custom-validators/video-captions.ts:12:88 - error TS2503: Cannot find namespace 'Express'.

12 function isVideoCaptionFile (files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[], field: string) {
                                                                                          ~~~~~~~

server/helpers/custom-validators/video-imports.ts:30:68 - error TS2503: Cannot find namespace 'Express'.

30 function isVideoImportTorrentFile (files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[]) {
                                                                      ~~~~~~~

server/helpers/custom-validators/video-imports.ts:30:94 - error TS2503: Cannot find namespace 'Express'.

30 function isVideoImportTorrentFile (files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[]) {
                                                                                                ~~~~~~~

server/lib/moderation.ts:40:14 - error TS2503: Cannot find namespace 'Express'.

40   videoFile: Express.Multer.File & { duration?: number }
                ~~~~~~~

server/middlewares/validators/videos/videos.ts:71:22 - error TS2503: Cannot find namespace 'Express'.

71     const videoFile: Express.Multer.File & { duration?: number } = req.files['videofile'][0]
                        ~~~~~~~

server/middlewares/validators/videos/videos.ts:482:89 - error TS2503: Cannot find namespace 'Express'.

482 async function isVideoAccepted (req: express.Request, res: express.Response, videoFile: Express.Multer.File & { duration?: number }) {
                                                                                            ~~~~~~~

server/helpers/express-utils.ts:32:64 - error TS2503: Cannot find namespace 'Express'.

32 function cleanUpReqFiles (req: { files: { [fieldname: string]: Express.Multer.File[] } | Express.Multer.File[] }) {
                                                                  ~~~~~~~

server/helpers/express-utils.ts:32:90 - error TS2503: Cannot find namespace 'Express'.

32 function cleanUpReqFiles (req: { files: { [fieldname: string]: Express.Multer.File[] } | Express.Multer.File[] }) {
                                                                                            ~~~~~~~

server/helpers/express-utils.ts:38:15 - error TS2503: Cannot find namespace 'Express'.

38     (files as Express.Multer.File[]).forEach(f => deleteFileAsync(f.path))
                 ~~~~~~~

server/typings/express/index.d.ts:41:16 - error TS2665: Invalid module name in augmentation. Module 'express' resolves to an untyped module at '/app/node_modules/express/index.js', which cannot be augmented.

41 declare module 'express' {
                  ~~~~~~~~~


Found 45 errors.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! peertube@2.4.0 tsc: `tsc "--build" "./server/tools/tsconfig.json"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the peertube@2.4.0 tsc script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2020-10-24T17_08_57_875Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! peertube@2.4.0 setup:cli: `scripty`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the peertube@2.4.0 setup:cli script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2020-10-24T17_08_57_895Z-debug.log
```

The first errors are obvious enough for me, I can install chai and supertest manually:

```
# yarn add chai supertest
# NOCLIENT=1 npm run setup:cli

> peertube@2.4.0 setup:cli /app
> scripty

yarn install v1.22.4
[1/5] Validating package.json...
[2/5] Resolving packages...
warning Resolution field "http-signature@1.3.4" is incompatible with requested version "http-signature@~1.2.0"
warning Resolution field "oauth2-server@3.1.0-beta.1" is incompatible with requested version "oauth2-server@3.0.0"
success Already up-to-date.
$ test -n "$NOCLIENT" || (cd client && yarn install --pure-lockfile)
Done in 0.62s.
yarn install v1.22.4
[1/4] Resolving packages...
success Already up-to-date.
Done in 0.29s.

> peertube@2.4.0 tsc /app
> tsc "--build" "./server/tools/tsconfig.json"

server/helpers/custom-validators/misc.ts:86:35 - error TS2503: Cannot find namespace 'Express'.

86   files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[],
                                     ~~~~~~~

server/helpers/custom-validators/misc.ts:86:61 - error TS2503: Cannot find namespace 'Express'.

86   files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[],
                                                               ~~~~~~~

server/helpers/custom-validators/users.ts:104:56 - error TS2503: Cannot find namespace 'Express'.

104 function isAvatarFile (files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[]) {
                                                           ~~~~~~~

server/helpers/custom-validators/users.ts:104:82 - error TS2503: Cannot find namespace 'Express'.

104 function isAvatarFile (files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[]) {
                                                                                     ~~~~~~~

server/helpers/custom-validators/videos.ts:83:55 - error TS2503: Cannot find namespace 'Express'.

83 function isVideoFile (files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[]) {
                                                         ~~~~~~~

server/helpers/custom-validators/videos.ts:83:81 - error TS2503: Cannot find namespace 'Express'.

83 function isVideoFile (files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[]) {
                                                                                   ~~~~~~~

server/helpers/custom-validators/videos.ts:92:56 - error TS2503: Cannot find namespace 'Express'.

92 function isVideoImage (files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[], field: string) {
                                                          ~~~~~~~

server/helpers/custom-validators/videos.ts:92:82 - error TS2503: Cannot find namespace 'Express'.

92 function isVideoImage (files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[], field: string) {
                                                                                    ~~~~~~~

server/helpers/custom-validators/video-captions.ts:12:62 - error TS2503: Cannot find namespace 'Express'.

12 function isVideoCaptionFile (files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[], field: string) {
                                                                ~~~~~~~

server/helpers/custom-validators/video-captions.ts:12:88 - error TS2503: Cannot find namespace 'Express'.

12 function isVideoCaptionFile (files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[], field: string) {
                                                                                          ~~~~~~~

server/helpers/custom-validators/video-imports.ts:30:68 - error TS2503: Cannot find namespace 'Express'.

30 function isVideoImportTorrentFile (files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[]) {
                                                                      ~~~~~~~

server/helpers/custom-validators/video-imports.ts:30:94 - error TS2503: Cannot find namespace 'Express'.

30 function isVideoImportTorrentFile (files: { [ fieldname: string ]: Express.Multer.File[] } | Express.Multer.File[]) {
                                                                                                ~~~~~~~

server/lib/moderation.ts:40:14 - error TS2503: Cannot find namespace 'Express'.

40   videoFile: Express.Multer.File & { duration?: number }
                ~~~~~~~

server/middlewares/validators/videos/videos.ts:71:22 - error TS2503: Cannot find namespace 'Express'.

71     const videoFile: Express.Multer.File & { duration?: number } = req.files['videofile'][0]
                        ~~~~~~~

server/middlewares/validators/videos/videos.ts:482:89 - error TS2503: Cannot find namespace 'Express'.

482 async function isVideoAccepted (req: express.Request, res: express.Response, videoFile: Express.Multer.File & { duration?: number }) {
                                                                                            ~~~~~~~

server/helpers/express-utils.ts:32:64 - error TS2503: Cannot find namespace 'Express'.

32 function cleanUpReqFiles (req: { files: { [fieldname: string]: Express.Multer.File[] } | Express.Multer.File[] }) {
                                                                  ~~~~~~~

server/helpers/express-utils.ts:32:90 - error TS2503: Cannot find namespace 'Express'.

32 function cleanUpReqFiles (req: { files: { [fieldname: string]: Express.Multer.File[] } | Express.Multer.File[] }) {
                                                                                            ~~~~~~~

server/helpers/express-utils.ts:38:15 - error TS2503: Cannot find namespace 'Express'.

38     (files as Express.Multer.File[]).forEach(f => deleteFileAsync(f.path))
                 ~~~~~~~

server/typings/express/index.d.ts:41:16 - error TS2665: Invalid module name in augmentation. Module 'express' resolves to an untyped module at '/app/node_modules/express/index.js', which cannot be augmented.

41 declare module 'express' {
                  ~~~~~~~~~


Found 19 errors.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! peertube@2.4.0 tsc: `tsc "--build" "./server/tools/tsconfig.json"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the peertube@2.4.0 tsc script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2020-10-24T17_02_04_909Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! peertube@2.4.0 setup:cli: `scripty`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the peertube@2.4.0 setup:cli script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2020-10-24T17_10_04_929Z-debug.log
```

The Express stuff, IDK.
I've looked at https://stackoverflow.com/questions/42388217/having-error-module-name-resolves-to-an-untyped-module-at-when-writing-cu
Couldn't figure it out yet ...

**What do you expect to happen instead?**

PeerTube CLI installed.

**Steps to reproduce:**

1. git clone (v2.4.0 or develop branch produces the same output)

2. follow instructions from https://framagit.org/framasoft/peertube/PeerTube/-/blob/develop/support/doc/tools.md (yarn install, npm run build)

**Comment by Chocobozzz** [comment]
Try in a fresh cloned PeerTube directory:

```
NOCLIENT=1 yarn install --pure-lockfile --production=false && npm run setup:cli
```

**Comment by faust64** [comment]
Ok, so I'ld start with:

```
FROM node:12.18.4-buster

ENV DEBIAN_FRONTEND=noninteractive \
    NODE_ENV=production \
    PATH="/app/.node-live/lib/node_modules/.bin:/app/.node-live/bin:$PATH"

USER root

RUN set -ex \
    && rm -rf /var/lib/apt/lists/* \
    && apt-get -y update \
    && echo "# Install PeerTube dependencies" \
    && apt-get -y install ldap-utils ca-certificates make git \
	g++ gnupg2 openssl postgresql-client ffmpeg \
    && mkdir -p /app \
    && echo "# Install PeerTube" \
    && git clone https://github.com/Chocobozzz/PeerTube /app \
    && cd /app \
    && NOCLIENT=1 yarn install --pure-lockfile  --production=false \
    && npm run setup:cli
```

This fails, as I'm still missing typescript:

```
> peertube@2.4.0 tsc /app
> tsc "--build" "./server/tools/tsconfig.json"

sh: 1: tsc: not found
npm ERR! code ELIFECYCLE
npm ERR! syscall spawn
npm ERR! file sh
npm ERR! errno ENOENT
npm ERR! peertube@2.4.0 tsc: `tsc "--build" "./server/tools/tsconfig.json"`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the peertube@2.4.0 tsc script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

I would then add some `npm install -g typescript` before the `setup:cli`.
Which also fails:

```
> peertube@2.4.0 tsc /app
> tsc "--build" "./server/tools/tsconfig.json"

shared/extra-utils/miscs/miscs.ts(3,23): error TS2307: Cannot find module 'chai' or its corresponding type declarations.
shared/extra-utils/miscs/miscs.ts(5,26): error TS2307: Cannot find module 'supertest' or its corresponding type declarations.
shared/extra-utils/requests/requests.ts(3,26): error TS2307: Cannot find module 'supertest' or its corresponding type declarations.
shared/extra-utils/server/servers.ts(7,24): error TS2307: Cannot find module 'chai' or its corresponding type declarations.
shared/extra-utils/server/clients.ts(1,26): error TS2307: Cannot find module 'supertest' or its corresponding type declarations.
shared/extra-utils/users/login.ts(1,26): error TS2307: Cannot find module 'supertest' or its corresponding type declarations.
shared/extra-utils/users/users.ts(1,26): error TS2307: Cannot find module 'supertest' or its corresponding type declarations.
shared/extra-utils/videos/video-channels.ts(3,26): error TS2307: Cannot find module 'supertest' or its corresponding type declarations.
server/helpers/custom-validators/misc.ts(86,35): error TS2503: Cannot find namespace 'Express'.
server/helpers/custom-validators/misc.ts(86,61): error TS2503: Cannot find namespace 'Express'.
server/helpers/custom-validators/users.ts(104,56): error TS2503: Cannot find namespace 'Express'.
server/helpers/custom-validators/users.ts(104,82): error TS2503: Cannot find namespace 'Express'.
shared/extra-utils/server/activitypub.ts(1,26): error TS2307: Cannot find module 'supertest' or its corresponding type declarations.
shared/extra-utils/server/jobs.ts(1,26): error TS2307: Cannot find module 'supertest' or its corresponding type declarations.
shared/extra-utils/server/follows.ts(1,26): error TS2307: Cannot find module 'supertest' or its corresponding type declarations.
shared/extra-utils/videos/services.ts(1,26): error TS2307: Cannot find module 'supertest' or its corresponding type declarations.
shared/extra-utils/videos/videos.ts(3,24): error TS2307: Cannot find module 'chai' or its corresponding type declarations.
shared/extra-utils/videos/videos.ts(7,26): error TS2307: Cannot find module 'supertest' or its corresponding type declarations.
shared/extra-utils/videos/video-playlists.ts(11,24): error TS2307: Cannot find module 'chai' or its corresponding type declarations.
shared/extra-utils/users/accounts.ts(3,26): error TS2307: Cannot find module 'supertest' or its corresponding type declarations.
shared/extra-utils/users/accounts.ts(4,24): error TS2307: Cannot find module 'chai' or its corresponding type declarations.
shared/extra-utils/videos/video-abuses.ts(1,26): error TS2307: Cannot find module 'supertest' or its corresponding type declarations.
shared/extra-utils/videos/video-blacklist.ts(1,26): error TS2307: Cannot find module 'supertest' or its corresponding type declarations.
shared/extra-utils/videos/video-captions.ts(2,26): error TS2307: Cannot find module 'supertest' or its corresponding type declarations.
shared/extra-utils/videos/video-captions.ts(3,23): error TS2307: Cannot find module 'chai' or its corresponding type declarations.
shared/extra-utils/videos/video-comments.ts(3,26): error TS2307: Cannot find module 'supertest' or its corresponding type declarations.
shared/extra-utils/videos/video-streaming-playlists.ts(4,24): error TS2307: Cannot find module 'chai' or its corresponding type declarations.
shared/extra-utils/videos/video-change-ownership.ts(1,26): error TS2307: Cannot find module 'supertest' or its corresponding type declarations.
shared/extra-utils/feeds/feeds.ts(1,26): error TS2307: Cannot find module 'supertest' or its corresponding type declarations.
shared/extra-utils/search/videos.ts(3,26): error TS2307: Cannot find module 'supertest' or its corresponding type declarations.
server/helpers/custom-validators/videos.ts(83,55): error TS2503: Cannot find namespace 'Express'.
server/helpers/custom-validators/videos.ts(83,81): error TS2503: Cannot find namespace 'Express'.
server/helpers/custom-validators/videos.ts(92,56): error TS2503: Cannot find namespace 'Express'.
server/helpers/custom-validators/videos.ts(92,82): error TS2503: Cannot find namespace 'Express'.
server/helpers/custom-validators/video-captions.ts(12,62): error TS2503: Cannot find namespace 'Express'.
server/helpers/custom-validators/video-captions.ts(12,88): error TS2503: Cannot find namespace 'Express'.
server/helpers/custom-validators/video-imports.ts(30,68): error TS2503: Cannot find namespace 'Express'.
server/helpers/custom-validators/video-imports.ts(30,94): error TS2503: Cannot find namespace 'Express'.
server/lib/moderation.ts(40,14): error TS2503: Cannot find namespace 'Express'.
server/middlewares/validators/videos/videos.ts(71,22): error TS2503: Cannot find namespace 'Express'.
server/middlewares/validators/videos/videos.ts(482,89): error TS2503: Cannot find namespace 'Express'.
server/helpers/express-utils.ts(32,64): error TS2503: Cannot find namespace 'Express'.
server/helpers/express-utils.ts(32,90): error TS2503: Cannot find namespace 'Express'.
server/helpers/express-utils.ts(38,15): error TS2503: Cannot find namespace 'Express'.
server/typings/express/index.d.ts(41,16): error TS2665: Invalid module name in augmentation. Module 'express' resolves to an untyped module at '/app/node_modules/express/index.js', which cannot be augmented.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! peertube@2.4.0 tsc: `tsc "--build" "./server/tools/tsconfig.json"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the peertube@2.4.0 tsc script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2020-10-31T07_05_55_988Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! peertube@2.4.0 setup:cli: `scripty`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the peertube@2.4.0 setup:cli script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

At which point I'ld install chai and supertest, fixing some of those errors. Yet I'm still unclear how to address the express ones.

Am I doing something wrong? Wrong typescript version maybe?

**Comment by rigelk** [comment]
@faust64 FYI we already have a Dockerfile in `support/doc/docker/production`.

**Comment by faust64** [comment]
I've seen. As well as that in `support/docker/production`, which is not bad.
I'ld rather build my own though, without gosu, latest nodejs runtime, adding nsswrapper/allowing for dynamic user ids (kubernetes/openshift support), additional scripts automating site provisioning, ...

But you're right, client builds properly, when using your Dockerfile.
So I started looking for an explanation, ...
Turns out I need to unset my `NODE_ENV=production`, for the `setup:cli`, and yarn install (the `--production=false` option alone wasn't enough)

So, I got it working, with:

```
ENV NODE_ENV=production
[...]
	&& NODE_ENV= yarn install --pure-lockfile --production=false \
	&& NODE_ENV= npm run build:server \
	&& NODE_ENV= npm run build:client \
	&& NODE_ENV= npm run setup:cli \
	&& rm -fr node_modules client/node_modules \
        && yarn install --pure-lockfile \
	&& npm cache clear --force; \
```

Though I could have moved that `ENV` in a layer following the build commands, as it is done in your production image, ... My mistake. Thanks!
